rlp/sedes/serializable.py

```diff
@@ -0,0 +1,299 @@
+import abc
+import collections
+
+from eth_utils import (
+    to_tuple,
+)
+
+from rlp.exceptions import (
+    ListSerializationError,
+    ObjectSerializationError,
+    ListDeserializationError,
+    ObjectDeserializationError,
+)
+
+from .lists import (
+    List,
+)
+
+
+class FieldsMeta:
+    fields = None
+    field_names = None
+    field_attrs = None
+    sedes = None
+
+
+@to_tuple
+def merge_args_and_kwargs(args, kwargs, arg_name_ordering):
+    if len(arg_name_ordering) != len(set(arg_name_ordering)):
+        raise TypeError("duplicate argument names")
+
+    needed_kwargs = arg_name_ordering[len(args):]
+    used_kwargs = set(arg_name_ordering[:len(args)])
+
+    duplicate_kwargs = used_kwargs.intersection(kwargs.keys())
+    if duplicate_kwargs:
+        raise TypeError("Duplicate kwargs: {0}".format(sorted(duplicate_kwargs)))
+
+    unknown_kwargs = set(kwargs.keys()).difference(arg_name_ordering)
+    if unknown_kwargs:
+        raise TypeError("Unknown kwargs: {0}".format(sorted(unknown_kwargs)))
+
+    missing_kwargs = set(needed_kwargs).difference(kwargs.keys())
+    if missing_kwargs:
+        raise TypeError("Missing kwargs: {0}".format(sorted(missing_kwargs)))
+
+    yield from args
+    for arg_name in needed_kwargs:
+        yield kwargs[arg_name]
+
+
+class BaseSerializable(collections.Sequence):
+    @property
+    @abc.abstractmethod
+    def _cached_rlp(self):
+        pass
+
+    def __init__(self, *args, **kwargs):
+        field_values = merge_args_and_kwargs(args, kwargs, self._meta.field_names)
+        if field_values:
+            field_attrs = self._get_field_attrs()
+
+            if len(field_values) != len(field_attrs):
+                raise TypeError('Argument count mismatch. expected {0} - got {1}'.format(
+                    len(field_attrs),
+                    len(field_values),
+                ))
+            for value, attr in zip(field_values, field_attrs):
+                setattr(self, attr, value)
+
+    def __iter__(self):
+        for field in self._meta.get_field_names():
+            yield getattr(self, field)
+
+    def __getitem__(self, idx):
+        if isinstance(idx, int):
+            field = self._meta.field_names[idx]
+            return getattr(self, field)
+        else:
+            raise NotImplementedError("not implemented")
+
+    def __len__(self):
+        return len(self._meta.fields)
+
+    def __eq__(self, other):
+        if not hasattr(other, 'serialize'):
+            return False
+        return self.serialize(self) == other.serialize(other)
```

Comment:
I'm not sure this is enough. An rlp with type (bytes32, bytes32) and one with type (int, int) can be trivially made to be equal here, which feels wrong to me.

Maybe test the deserialized values?

Reply:
Agreed.  

> Note that this is the current way that `rlp.sedes.Serializable` determines equality: https://github.com/ethereum/pyrlp/blob/188abd6f536aaaf9689413650422991b7510f9a7/rlp/sedes/lists.py#L204-L208

I may go back to using a version of the following:

```python
return len(self) == len(other) and all(
    compare_eq(left, right)
    for left, right
    in zip(self, other)
)
```

Where `compare_eq` is a function which handles things like equality between `tuples` and `lists` with the same values which a naive `==` comparison will return `False`.